pandas/io/excel.py

```diff
@@ -66,27 +69,211 @@ def get_writer(engine_name):
+        if self.io_class is None:
+            return False
+        return ext in self.extensions
+
+    def load_engine(self):
+        """Load the engine if installed, but raise ImportError otherwise.
+        If engine is not defined, do nothing.
+        """
+        try:
+            _readers[self.engine] = self._load_engine()
+        except AttributeError:
+            _readers[self.engine] = self._load_engine()
+            msg = 'Excel reader engine "%s" is not implemented' % self.engine
+            raise NotImplementedError(msg)
+
+    def _load_engine(self):
+        return False
+
+    def create(self, io, engine=None):
+        """Create the appropriate speadsheet book based on io and optionally
+        engine.
+
+        Paratemeters
+        ------------
+        io : string, file-like object or xlrd/ezodf workbook
+            If a string, expected to be a path to xls, xlsx, or ods file.
+            File-like objects or buffers are only supported for xlrd types.
+
+        engine: string, default None
+            If io is not a buffer or path, this must be set to identify io.
+            Acceptable values are None, xlrd, or ezodf
+
+        Returns
+        -------
+        engine : string
+            Engine used for reading the io
+
+        book : object
+            The spreadsheet book class created by the engine.
+        """
+        self.engine = engine
+        # Initiate pandas supported file readers. When creating new readers,
+        # add them to the supported_readers list
+        xlrd_f = XLRDFile()
+        ezodf_f = EZODFFile()
+        supported_readers = [xlrd_f, ezodf_f]
+
+        for reader in supported_readers:
+            if self.engine == reader.engine:
+                self.book = reader.open_workbook(io)
+                return self.engine, self.book
+            # derive engine from file extension if io is a path/url
+            elif isinstance(io, compat.string_types):
+                if reader.is_ext(io):
+                    self.engine = reader.engine
+                    if _is_url(io):
+                        data = _urlopen(io).read()
+                        # open_workbook: BaseFile will raise ImportError
+                        # when the engine is not installed
+                        self.book = reader.read_buffer(data)
+                    else:
+                        self.book = reader.open_workbook(io)
+                    return self.engine, self.book
+            # Does the io type match any supported and installed reader types?
+            # Note that io_class is None if engine is not installed
+            elif reader.io_class is not None and isinstance(io, reader.io_class):
+                self.engine = reader.engine
+                self.book = io
+                return self.engine, self.book
+
+        # xlrd has some additional/alternative reading mechanisms:
+        if not isinstance(io, xlrd_f.io_class) and hasattr(io, "read"):
```

Comment:
this can be encapsulated in a method for each reader, e.g. in the xlrd, ezodf sub-classes, e.g. something like
hen the loop becomes

```
for reader in self.supported_readers:
    try:
          return reader.create_engine(io, engine=engine):
    except (NotImplementedError):
         pass

raise ValueError("no supported readers for this input")
```